docs/pages/router/reference/authentication.mdx

```diff
@@ -5,115 +5,352 @@ description: How to protect routes with Expo Router.
+      localStorage.removeItem(key);
+    } else {
+      localStorage.setItem(key, value);
+    }
+  } else {
+    if (value == null) {
+      await SecureStore.deleteItemAsync(key);
+    } else {
+      await SecureStore.setItemAsync(key, value);
+    }
+  }
+}
+
+export function useStorageState(key: string): UseStateHook<string> {
+  // Public
+  const [state, setState] = useAsyncState<string>();
+
+  // Get
+  React.useEffect(() => {
+    if (Platform.OS === 'web') {
+      if (typeof localStorage !== 'undefined') {
+        setState(localStorage.getItem(key));
+      }
+    } else {
+      SecureStore.getItemAsync(key).then(value => {
+        setState(value);
+      });
+    }
+  }, [key]);
+
+  // Set
+  const setValue = React.useCallback(
+    (value: string | null) => {
+      setStorageItemAsync(key, value).then(() => {
+        setState(value);
+      });
+    },
+    [key]
+  );
+
+  return [state, setValue];
+}
+```
+
+</Collapsible>
+
+Now we can use our `SessionProvider` in the root layout to provide the authentication context to the entire app. It's imperative that the `<Slot />` is mounted before any navigation events are triggered, otherwise a runtime error will be thrown.
 
 ```jsx app/_layout.js
 import { Slot } from 'expo-router';
-import { Provider } from '../context/auth';
+import { SessionProvider } from '../ctx';
 
 export default function Root() {
+  // Setup the auth context and render our layout inside of it.
   return (
-    // Setup the auth context and render our layout inside of it.
-    <Provider>
+    <SessionProvider>
       <Slot />
-    </Provider>
+    </SessionProvider>
   );
 }
 ```
 
-Now we can create our `(auth)` group which is unprotected, this screen can toggle the authentication using `signIn()`.
+Next, we'll create a nested [Layout Route](/routing/layouts) which will be used to protect all of the children. This layout will be used to redirect users to the sign in screen if they are not authenticated.
+
+```jsx app/(app)/_layout.js
+import { Link, Redirect, Stack } from 'expo-router';
+import { Text, View } from 'react-native';
+
+import { useSession } from '../../ctx';
+
+export default function AppLayout() {
+  const { session, isLoading } = useSession();
+
+  // You can keep the splash screen open, or render a loading screen.
+  if (isLoading) {
+    return <Text>Loading...</Text>;
+  }
+
+  // Only do authorization on the (app) group as users should always be able to access the (auth) group and sign-in again.
+  if (!session) {
+    // On web, static rendering will stop here as the user is not authenticated in the headless Node process that the pages are rendered in.
+    return <Redirect href="/sign-in" />;
+  }
+
+  // This layout can be deferred because it's not the root layout.
+  return <Stack />;
+}
+```
+
+Now we can create our `/sign-in` page, which can toggle the authentication using `signIn()`. Because this screen is outside of the `(app)` group, it won't be protected by the layout.
```

Comment:
```suggestion
Now we can create our `/sign-in` page, which can toggle the authentication using `signIn()`. Because this screen is outside of the `(app)` group, the `(app)` group's layout and its authentication check are not run when rendering this screen. This lets logged-out users see this screen.
```